packages/api-server/src/schema/squiggle.ts

```diff
@@ -0,0 +1,160 @@
+import crypto from "crypto";
+
+import {
+  SqAbstractDistribution,
+  SqProject,
+  SqValue,
+} from "@quri/squiggle-lang";
+
+import { Prisma, SquiggleCache } from "@prisma/client";
+import { builder } from "../builder";
+import { prisma } from "../prisma";
+
+function getKey(code: string): string {
+  return crypto.createHash("md5").update(code).digest("base64");
+}
+
+const squiggleValueToJSON = (value: SqValue) => {
+  // this is a lazy shortcut to traverse the value tree; should be reimplemented without parse/stringify
+  return JSON.parse(
+    JSON.stringify(value.asJS(), (key, value) => {
+      if (value instanceof Map) {
+        return Object.fromEntries(value.entries());
+      }
+      if (value instanceof SqAbstractDistribution) {
+        return value.toString();
+      }
+      return value;
+    })
+  );
+};
+
+type SquiggleOutput = {
+  isCached: boolean;
+} & (
+  | {
+      isOk: false;
+      errorString: string;
+      resultJSON?: undefined;
+      bindingsJSON?: undefined;
+    }
+  | {
+      isOk: true;
+      errorString?: undefined;
+      resultJSON: Prisma.JsonValue;
+      bindingsJSON: Prisma.JsonValue;
+    }
+);
+
+function runSquiggle(code: string): SquiggleOutput {
+  const MAIN = "main";
+
+  const project = SqProject.create();
+
+  project.setSource(MAIN, code);
+  project.run(MAIN);
+
+  const result = project.getResult(MAIN);
+  const bindings = project.getBindings(MAIN);
+
+  return result.ok
+    ? {
+        isCached: false,
+        isOk: true,
+        resultJSON: squiggleValueToJSON(result.value),
+        bindingsJSON: squiggleValueToJSON(bindings.asValue()),
+      }
+    : {
+        isCached: false,
+        isOk: false,
+        errorString: result.value.toString(),
+      };
+}
+
+const SquiggleOutputObj = builder
+  .interfaceRef<SquiggleOutput>("SquiggleOutput")
+  .implement({
+    fields: (t) => ({
+      isCached: t.exposeBoolean("isCached"),
+    }),
+  });
+
+builder.objectType(
+  builder.objectRef<Extract<SquiggleOutput, { isOk: true }>>(
+    "SquiggleOkOutput"
+  ),
+  {
+    name: "SquiggleOkOutput",
+    interfaces: [SquiggleOutputObj],
+    isTypeOf: (value) => (value as SquiggleOutput).isOk,
+    fields: (t) => ({
+      resultJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.resultJSON);
+        },
+      }),
+      bindingsJSON: t.string({
+        resolve(obj) {
+          return JSON.stringify(obj.bindingsJSON);
+        },
```

Comment:
I find it interesting you're saving bindings too. Curious what we can later do with that.